### ISSUE TITLE
With new yield strategy, display_end hook is unusable.

### ISSUE THREAD

**Issue by TLG-Gildas** [twigphp/Twig#4008]
With new yield strategy the display_end node can't be reach when template doesn't have parent.

Indeed, in this case, the node is compiled after a `return` and therefore the code produced can never be executed.

See https://github.com/twigphp/Twig/blob/b46e93c7257fb01b7c77768210997b1e00643b91/src/Node/ModuleNode.php#L316-L347

especially the line
https://github.com/twigphp/Twig/blob/b46e93c7257fb01b7c77768210997b1e00643b91/src/Node/ModuleNode.php#L340
which is generated when template doesn't have parent.
and the line https://github.com/twigphp/Twig/blob/b46e93c7257fb01b7c77768210997b1e00643b91/src/Node/ModuleNode.php#L344 which is generated after the `return`

display_end node must be reachable.

**Comment by andypost** [comment]
It blocking adoption in [Drupal](https://www.drupal.org/project/drupal/issues/3441331)  which is using 

```php

class TwigNodeVisitorCheckDeprecations implements NodeVisitorInterface {
  public function leaveNode(Node $node, Environment $env): ?Node {
    // At the end of the template, check the used variables are not deprecated.
    if ($node instanceof ModuleNode) {
      if (!empty($this->usedNames)) {
        $checkNode = new Node([new TwigNodeCheckDeprecations($this->usedNames), $node->getNode('display_end')]);
        $node->setNode('display_end', $checkNode);
      }
    }
    return $node;
  }
```

**Comment by ericmorand** [comment]
@andypost why is Drupal adding some linting logic to their runtime framework? Code quality has already been processed at development time, why add it at execution time?

**Comment by stof** [comment]
In any case, we should not break support for the `display_end` extension point.

**Comment by fabpot** [comment]
@TLG-Gildas @andypost Can you confirm that #4028 fixes the issue?

**Comment by longwave** [comment]
Confirmed that Drupal's TwigDeprecationsTest passes with #4028 where it did not before.

**Comment by TLG-Gildas** [comment]
> @TLG-Gildas @andypost Can you confirm that #4028 fixes the issue?

same on my side #4028 solves the issue

**Comment by andypost** [comment]
@ericmorand Drupal has sophisticated template override system so we can't be sure all contrib and even custom code is updated and using proper linters at all. So render is the only way to make sure template is using deprecated variable. The whole story is https://www.drupal.org/node/3270148

**Comment by fabpot** [comment]
I've just released 3.9.2 which includes the fix for this issue.
